readthedocs/oauth/services/base.py

```diff
@@ -149,96 +146,14 @@ def for_project(cls, project):
             yield cls(user=user, account=account)
 
-    def get_adapter(self) -> type[OAuth2Adapter]:
-        return self.adapter
-
-    @property
-    def provider_id(self):
-        return self.get_adapter().provider_id
-
-    def get_session(self):
-        if self.session is None:
-            self.create_session()
-        return self.session
-
-    def get_access_token_url(self):
-        # ``access_token_url`` is a property in some adapters,
-        # so we need to instantiate it to get the actual value.
-        # pylint doesn't recognize that get_adapter returns a class.
-        # pylint: disable=not-callable
-        adapter = self.get_adapter()(request=None)
-        return adapter.access_token_url
-
-    def create_session(self):
-        """
-        Create OAuth session for user.
-
-        This configures the OAuth session based on the :py:class:`SocialToken`
-        attributes. If there is an ``expires_at``, treat the session as an auto
-        renewing token. Some providers expire tokens after as little as 2 hours.
-        """
-        token = self.account.socialtoken_set.first()
-        if token is None:
-            return None
-
-        token_config = {
-            "access_token": token.token,
-            "token_type": "bearer",
-        }
-        if token.expires_at is not None:
-            token_expires = (token.expires_at - timezone.now()).total_seconds()
-            token_config.update(
-                {
-                    "refresh_token": token.token_secret,
-                    "expires_in": token_expires,
-                }
-            )
-
-        social_app = self.account.get_provider().app
-        self.session = OAuth2Session(
-            client_id=social_app.client_id,
-            token=token_config,
-            auto_refresh_kwargs={
-                "client_id": social_app.client_id,
-                "client_secret": social_app.secret,
-            },
-            auto_refresh_url=self.get_access_token_url(),
-            token_updater=self.token_updater(token),
-        )
-
-        return self.session or None
-
-    def token_updater(self, token):
-        """
-        Update token given data from OAuth response.
-
-        Expect the following response into the closure::
-
-            {
-                u'token_type': u'bearer',
-                u'scopes': u'webhook repository team account',
-                u'refresh_token': u'...',
-                u'access_token': u'...',
-                u'expires_in': 3600,
-                u'expires_at': 1449218652.558185
-            }
-        """
-
-        def _updater(data):
-            token.token = data["access_token"]
-            token.token_secret = data.get("refresh_token", "")
-            token.expires_at = timezone.make_aware(
-                datetime.fromtimestamp(data["expires_at"]),
-            )
-            token.save()
-            log.info("Updated token.", token_id=token.pk)
-
-        return _updater
+    @cached_property
```

Comment:
What is caching doing here? Seems a little scary to cache an object like this? Could probably use a comment at least explaining why we aren't just setting this in the `__init__` as `self.session`?

Reply:
This is the same as we used to have (cache the client at the instance level), but we were manually caching the object.

https://github.com/readthedocs/readthedocs.org/blob/2c963f156e789bede74e36ae7a76b43f1918fdf6/readthedocs/oauth/services/base.py#L164-L167